### PR TITLE
Add ERC: Basket Token

### PR DESCRIPTION
**Pull request by AlvaraProtocol**
Initial ERC-BTS draft submission for purpose of discussion in forum.

### CODE REVIEW

File: ERCS/eip-draft-erc-bts.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+---
+title: ERC-BTS (Basket Token Standard)
```
Comment (abcoathup) [inline ERCS/eip-draft-erc-bts.md:2]:
```suggestion
title: Basket Token
```
You don't need to include ERC in the title.  
It is suggested not to include the word standard, this is implied by being an ERC.

File: ERCS/eip-draft-erc-bts.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
+---
```
Comment (abcoathup) [inline ERCS/eip-draft-erc-bts.md:1]:
```suggestion
---
eip: 7621
```
Assigning next sequential EIP/ERC/RIP number.
Numbering changed to sequential from 7500 and is no longer the PR number.

Please also update the filename.

File: ERCS/erc-7621.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+withdraw(uint256 _liquidity): Allows BTS LP holders to withdraw from the basket, receiving corresponding tokens.
+
+withdrawETH(uint256 _liquidity): Allows BTS LP holders to withdraw ETH from the basket, equivalent to the value of their BTS LP tokens.
+
+rebalance(address[] memory _newTokens, uint256[] memory _newWeights): The owner can manually adjust the types and weights of tokens in the basket.
+
+getTokenDetails: Returns details of a token at a given index (address and weight).
+
+totalTokens: Returns the total number of tokens in the basket.
```
Comment (gaudren) [inline ERCS/erc-7621.md:50]:
```suggestion
`initialize`: Initializes a new BTS with name, symbol, creator, tokens, weights, token URI, and optional auto-rebalance setting.

`contribute()`: Allows users to add ETH to the basket, purchasing proportional amounts of included tokens based on predefined weights.

`withdraw(uint256 _liquidity)`: Allows BTS LP holders to withdraw from the basket, receiving corresponding tokens.

`withdrawETH(uint256 _liquidity)`: Allows BTS LP holders to withdraw ETH from the basket, equivalent to the value of their BTS LP tokens.

`rebalance(address[] memory _newTokens, uint256[] memory _newWeights)`: The owner can manually adjust the types and weights of tokens in the basket.

`getTokenDetails`: Returns details of a token at a given index (address and weight).

`totalTokens`: Returns the total number of tokens in the basket.
```

Code snippets should be in backticks.

File: ERCS/erc-7621.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+#### Distribution of BTS tokens
+
+The distribution of BTS tokens occurs during the mint function. The creator sends ETH to the contract, and the contract then swaps the ETH for user designated [ERC-20](./eip-20.md) tokens based on predefined weights. After the swaps, it mints a BTS token (NFT) for the sender using the initialize function of the BTS contract as well as BTS LP tokens. This BTS token is a representation of ownership and allows management of the BTS.
+
+#### Distribution of BTS LP tokens
+
+The distribution of BTS LP tokens occurs during the mint and contribute functions. After swapping ETH for the relevant [ERC-20](./eip-20.md) tokens, the contract mints BTS LP tokens (representing the user's share of the total BTS), and then mints BTS LP tokens using the mint function of the IBTSPair contract. The BTS LP tokens represent the liquidity provided by the user to the specified BTS pair (btsPair). The distribution is logged through the ContributedToBTS event.
```
Comment (gaudren) [inline ERCS/erc-7621.md:58]: > IBTSPair

Should this be BTSPair?

Comment (gaudren) [inline ERCS/erc-7621.md:58]:
```suggestion
The distribution of BTS LP tokens occurs during the mint and contribute functions. After swapping ETH for the relevant [ERC-20](./eip-20.md) tokens, the contract mints BTS LP tokens (representing the user's share of the total BTS), and then mints BTS LP tokens using the mint function of the IBTSPair contract. The BTS LP tokens represent the liquidity provided by the user to the specified BTS pair (`btsPair`). The distribution is logged through the `ContributedToBTS` event.
```

File: ERCS/erc-7621.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+#### Burning the BTS LP tokens
+
+The burning of BTS LP tokens occurs during the withdraw function. Users can call this function, specifying the amount of BTS LP tokens they want to withdraw. The contract then transfers the specified amount of BTS LP tokens from the user to the BTS pair contract (btsPair). The burn function of the IBTSPair contract is called, which decreases the user's BTS LP token balance and returns an array of amounts representing the underlying tokens withdrawn. This array of amounts is logged through the WithdrawnFromBTS event.
```
Comment (gaudren) [inline ERCS/erc-7621.md:62]:
```suggestion
The burning of BTS LP tokens occurs during the `withdraw` function. Users can call this function, specifying the amount of BTS LP tokens they want to withdraw. The contract then transfers the specified amount of BTS LP tokens from the user to the BTS pair contract (`btsPair`). The burn function of the `IBTSPair` contract is called, which decreases the user's BTS LP token balance and returns an array of amounts representing the underlying tokens withdrawn. This array of amounts is logged through the `WithdrawnFromBTS` event.
```

File: ERCS/erc-7621.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+#### Events
+
+ContributedToBTS(address indexed bts, uint256 amount): Event when someone adds ETH to the basket.
+
+WithdrawnFromBTS(address indexed bts, uint256[] amounts): Event when BTS LP holder withdraws tokens from the basket.
+
+WithdrawnETHFromBTS(address indexed bts, uint256 amount): Event when BTS LP holder withdraws ETH from the basket.
+
+RebalanceBTS(address indexed bts, uint256[] oldWeights, uint256[] newWeights): Event when the owner manually rebalances the basket.
```
Comment (gaudren) [inline ERCS/erc-7621.md:72]:
```suggestion
`ContributedToBTS(address indexed bts, uint256 amount)`: Event when someone adds ETH to the basket.

`WithdrawnFromBTS(address indexed bts, uint256[] amounts)`: Event when BTS LP holder withdraws tokens from the basket.

`WithdrawnETHFromBTS(address indexed bts, uint256 amount)`: Event when BTS LP holder withdraws ETH from the basket.

`RebalanceBTS(address indexed bts, uint256[] oldWeights, uint256[] newWeights)`: Event when the owner manually rebalances the basket.
```

File: ERCS/erc-7621.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+btsPair: A public variable storing the address of the associated BTS pair contract.
+
+autoRebalanceEnabled: A public variable indicating whether auto-rebalancing is enabled or not.
+
+mapping(address => bool) private _isTokenPresent: A private mapping indicating whether a token is present or not.
+
+TokenDetails: A structure holding details about the tokens, including arrays for token addresses and their corresponding weights.
+
+_tokenDetails: An instance of the TokenDetails structure.
+
```
Comment (gaudren) [inline ERCS/erc-7621.md:85]:
```suggestion

```

This looks like implementation details. Contracts should be free to use whatever storage layout they want.

File: ERCS/erc-7621.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+### BTSPair
```
Comment (gaudren) [inline ERCS/erc-7621.md:86]:
```suggestion
### `BTSPair`
```

File: ERCS/erc-7621.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+By leveraging the [ERC-721](./eip-721.md) standard as a representative token of ownership when minting the BTS, the tokenized fund can also be fully manageable and transferable on-chain.
+
+## Motivation
+
+The motivation is to provide infrastructure that will enable the on-chain creation and management of asset-backed tokenized investment funds.
+
+## Specification
+
+### BTS
```
Comment (gaudren) [inline ERCS/erc-7621.md:30]: Write some introductory text. It is not clear that BTS is a contract.

File: ERCS/erc-7621.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+initialize: Initializes a new BTSPair with a name and tokens.
+
+rebalance(): Allows the owner to manually adjust the types and weights of tokens in the pair.
+
+updateTokens(...): Allows the owner to change the types of tokens in the pair.
+
+mint(...): Creates new BTS LP tokens and adjusts token reserves.
+
+burn(...): Destroys BTS LP tokens and adjusts token reserves.
```
Comment (gaudren) [inline ERCS/erc-7621.md:102]:
```suggestion
`initialize`: Initializes a new `BTSPair` with a name and tokens.

`rebalance()`: Allows the owner to manually adjust the types and weights of tokens in the pair.

`updateTokens(...)`: Allows the owner to change the types of tokens in the pair.

`mint(...)`: Creates new BTS LP tokens and adjusts token reserves.

`burn(...)`: Destroys BTS LP tokens and adjusts token reserves.
```

File: ERCS/erc-7621.md [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+#### Storage
+
+tokens: Array of token addresses in the pair.
+
+reserves: Array of corresponding token reserves.
```
Comment (gaudren) [inline ERCS/erc-7621.md:108]:
```suggestion

```

File: ERCS/erc-7621.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+### IUniswap
```
Comment (gaudren) [inline ERCS/erc-7621.md:110]:
```suggestion
### `IUniswap`
```

File: ERCS/erc-7621.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+#### Purpose
+
+Defines interfaces for interacting with Uniswap V2 Router and Factory contracts (external dependencies).
+
+## Rationale
```
Comment (gaudren) [inline ERCS/erc-7621.md:116]: If you include an HTML style comment like `<!-- TODO -->`, the linter will remind you to fill in the section.

File: ERCS/erc-7621.md (hunk 12 above)
Comment (gaudren) [inline ERCS/erc-7621.md:110]: This section hasn't provided any useful information for implementors.

File: ERCS/erc-7621.md [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+## Reference Implementation
+
+BTS.sol: Implemented using a combination of Solidity's built-in [ERC-721](./eip-721.md)URIStorage and Ownable contracts, along with custom logic for token management and rebalancing.
+
+Factory.sol: Implemented using ClonesUpgradeable to deploy new BTS and BTSPair contracts.
+
+BTSPair.sol: Implemented using [ERC-20](./eip-20.md)Upgradeable and Ownable contracts, along with custom logic for liquidity tracking and rebalancing.
+
+IUniswap.sol: Not implemented, as it only defines interfaces for external contracts.
```
Comment (gaudren) [inline ERCS/erc-7621.md:132]: The solidity files should be linked to the `assets` directory.

File: ERCS/erc-7621.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+## Security Considerations
```
Comment (gaudren) [inline ERCS/erc-7621.md:134]: If you include an HTML style comment like `<!-- TODO -->`, the linter will remind you to fill in the section.

**Review by AlvaraProtocol — COMMENTED** [review]
All comments updated.

File: ERCS/erc-7621.md [hunk 16]
```diff
@@ -0,0 +1,118 @@
+---
+eip: 7621
+title: Basket Token
+description: A specification for the creation of collateralized, tokenized funds with adjustable weights and reserves and minting/burning of LP tokens.
```
Comment (gaudren) [inline ERCS/erc-7621.md:4]:
```suggestion
description: Collateralized, tokenized funds with adjustable weights and reserves and minting/burning of LP tokens.
```

File: ERCS/erc-7621.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+`totalTokens`: Returns the total number of tokens in the basket.
+
+#### Distribution of BTS tokens
+
+The distribution of BTS tokens occurs during the mint function. The creator sends ETH to the contract, and the contract then swaps the ETH for user designated [ERC-20](./eip-20.md) tokens based on predefined weights. After the swaps, it mints a BTS token (NFT) for the sender using the initialize function of the BTS contract as well as BTS LP tokens. This BTS token is a representation of ownership and allows management of the BTS.
+
+#### Distribution of BTS LP tokens
+
+The distribution of BTS LP tokens occurs during the mint and contribute functions. After swapping ETH for the relevant [ERC-20](./eip-20.md) tokens, the contract mints BTS LP tokens (representing the user's share of the total BTS), and then mints BTS LP tokens using the mint function of the BTSPair contract. The BTS LP tokens represent the liquidity provided by the user to the specified BTS pair (`btsPair`). The distribution is logged through the `ContributedToBTS` event.
```
Comment (gaudren) [inline ERCS/erc-7621.md:60]:
```suggestion
The distribution of BTS LP tokens occurs during the mint and contribute functions. After swapping ETH for the relevant [ERC-20](./eip-20.md) tokens, the contract mints BTS LP tokens (representing the user's share of the total BTS), and then mints BTS LP tokens using the mint function of the `BTSPair` contract. The BTS LP tokens represent the liquidity provided by the user to the specified BTS pair (`btsPair`). The distribution is logged through the `ContributedToBTS` event.
```

File: ERCS/erc-7621.md [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+<!-- TODO -->
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Reference Implementation
+
+BTS: Implemented using a combination of Solidity's built-in [ERC-721](./eip-721.md)URIStorage and Ownable contracts, along with custom logic for token management and rebalancing.
```
Comment (gaudren) [inline ERCS/erc-7621.md:104]: Does Solidity have a  built-in 721?

Comment (gaudren) [inline ERCS/erc-7621.md:104]:
```suggestion
`BTS`: Implemented using a combination of Solidity's built-in [ERC-721](./eip-721.md) `URIStorage` and `Ownable` contracts, along with custom logic for token management and rebalancing.
```

File: ERCS/erc-7621.md [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+Factory: Implemented using ClonesUpgradeable to deploy new BTS and BTSPair contracts.
```
Comment (gaudren) [inline ERCS/erc-7621.md:106]:
```suggestion
`Factory`: Implemented using `ClonesUpgradeable` to deploy new `BTS` and `BTSPair` contracts.
```

File: ERCS/erc-7621.md [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+BTSPair: Implemented using [ERC-20](./eip-20.md)Upgradeable and Ownable contracts, along with custom logic for liquidity tracking and rebalancing.
```
Comment (gaudren) [inline ERCS/erc-7621.md:108]:
```suggestion
`BTSPair`: Implemented using [ERC-20](./eip-20.md) `Upgradeable` and `Ownable` contracts, along with custom logic for liquidity tracking and rebalancing.
```

File: ERCS/erc-7621.md [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+
+IUniswap: Not implemented, as it only defines interfaces for external contracts.
```
Comment (gaudren) [inline ERCS/erc-7621.md:110]:
```suggestion
`IUniswap`: Not implemented, as it only defines interfaces for external contracts.
```

File: ERCS/erc-7621.md (hunk 18 above)
Comment (AlvaraProtocol) [reply ERCS/erc-7621.md]: Should read OpenZeppelin's ERC-721

**Review by gaudren — APPROVED** [review]

**Review by SamWilsn — APPROVED** [review]
Good to go as a draft. Just make sure to add your reference implementation code or remove that section before going into review.

**Pull request merged**